### PR TITLE
Hexagon: Vrmpy double vector instruction

### PR DESCRIPTION
**Pull request by aankit-ca**
Vrmpy double vector instruction for Hexagon target.

### CODE REVIEW

**Comment by aankit-ca** [comment]
@dsharletg PTAL

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -1698,23 +1698,39 @@ class VtmpyGenerator : public IRMutator {
...
 
+    // Increament the load index of the expression by 2 if
+    // the expr is if of type load otherwise return empty expr.
+    Expr shift_load_by_2(const Expr &e) {
+        Expr load = calc_load(e);
+        if (load.defined()) {
+            const Load *maybe_load = load.as<Load>();
+            if (maybe_load) {
+                return Load::make(maybe_load->type, maybe_load->name,
+                                maybe_load->index + make_two(maybe_load->index.type()),
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Indentation is off

File: src/HexagonOptimize.cpp [hunk 2]
```diff
@@ -1760,38 +1785,84 @@ class VtmpyGenerator : public IRMutator {
...
+                                }
+
+                                Expr shifted_expr = shift_load_by_2(mpys[v2_idx].first);
+
+                                sw_indices[v0_idx] = true;
+                                sw_indices[v1_idx] = true;
+                                sw_indices[v2_idx] = true;
+
+                                Expr b0123 = Shuffle::make_interleave({ mpys[v0_idx].second,
+                                                mpys[v1_idx].second, mpys[v2_idx].second, b3 });
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Funky indentation.

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -1760,38 +1785,84 @@ class VtmpyGenerator : public IRMutator {
...
+                                b0123 = simplify(b0123);
+                                b0123 = reinterpret(Type(b0123.type().code(), 32, 1), b0123);
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Might be better to just use Int(32) for clarity and to constrain the code. This has to be 4 x Int(8) right?

File: test/correctness/simd_op_check.cpp [hunk 4]
```diff
@@ -1889,6 +1889,9 @@ struct Test {
         check("v*:*.w = vtmpy(v*:*.h, r*.b)", hvx_width/2, i32(in_i16(x - 1))*2 + i32(in_i16(x)) + i32(in_i16(x + 1)));
         check("v*:*.w = vtmpy(v*:*.h, r*.b)", hvx_width/2, i32(in_i16(x - 1)) + i32(in_i16(x)) + i32(in_i16(x + 1)));
 
+        check("v*:*.w = vrmpy(v*:*.ub, r*.b, #*)", hvx_width/1, 2*i32(in_u8(x - 1)) + 3*i32(in_u8(x)) + 5*i32(in_u8(x + 1)) + 7*i32(in_u8(x + 2)));
+        check("v*:*.w = vrmpy(v*:*.ub, r*.b, #*)", hvx_width/1, 9*i32(in_u8(x - 1)) + 8*i32(in_u8(x)) + 7*i32(in_u8(x + 1)));
```
Comment (dsharletg) [inline test/correctness/simd_op_check.cpp]: What is the # syntax? Is this required to be an immediate value? If so, I think there should be some asserts in the code that generates vrmpy to make sure the scalars are constants.

Comment (aankit-ca) [inline test/correctness/simd_op_check.cpp]: Yes it is an immediate value {0,1}. The constant value is passed as an argument in the call at L1846, L1841 (https://github.com/halide/Halide/pull/2095/files#diff-007baef34b5299fdf08df3e32378726eR1846) with implementation for vrmpy in hvx_128.ll.

File: src/HexagonOptimize.cpp (hunk 3 above)
Comment (aankit-ca) [inline src/HexagonOptimize.cpp]: Changed

File: test/correctness/simd_op_check.cpp (hunk 4 above)
Comment (dsharletg) [inline test/correctness/simd_op_check.cpp]: Ah of course, I see. At first I thought that was the multiply operand, now I see that it's the control flag.

**Comment by aankit-ca** [comment]
@dsharletg PTAL

**Comment by ronlieb** [comment]
ping

**Comment by dsharletg** [comment]
Sorry for the delay, we're working on testing this, there are some unrelated issues we're working on first.

**Comment by steven-johnson** [comment]
This PR is pretty ancient. Is it still an ongoing concern?

**Comment by aankit-ca** [comment]
I am closing the PR as this does not seem much relevant now.

**Pull request closed without merging**